services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/user_services_preferences/_manager.py

```diff
@@ -0,0 +1,69 @@
+import logging
+from pathlib import Path
+
+from attr import dataclass
+from fastapi import FastAPI
+from models_library.products import ProductName
+from models_library.services import ServiceKey, ServiceVersion
+from models_library.users import UserID
+
+from . import _db
+from ._utils import is_feature_enabled
+
+_logger = logging.getLogger(__name__)
+
+
+@dataclass
+class UserServicesPreferencesManager:
+    user_preferences_path: Path
+    service_key: ServiceKey
+    service_version: ServiceVersion
+    user_id: UserID
+    product_name: ProductName
+    _preferences_already_saved: bool = False
+
+    async def load_preferences(self) -> None:
+        await _db.load_preferences(
+            user_preferences_path=self.user_preferences_path,
+            service_key=self.service_key,
+            service_version=self.service_version,
+            user_id=self.user_id,
+            product_name=self.product_name,
+        )
+
+    async def save_preferences(self) -> None:
+        if self._preferences_already_saved:
+            _logger.warning("Preferences were already saved, will not save them again")
+            return
+
+        await _db.save_preferences(
+            user_preferences_path=self.user_preferences_path,
+            service_key=self.service_key,
+            service_version=self.service_version,
+            user_id=self.user_id,
+            product_name=self.product_name,
+        )
+
+        self._preferences_already_saved = True
+
+
+async def save_user_services_preferences(app: FastAPI) -> None:
+    if not is_feature_enabled(app):
+        _logger.info("preferences features is disabled, skipping save")
```

Comment:
Thought: Do you think it is a good idea to call something like `save_user_services_preferences` and silently (typically we set LOG_LEVEL=WARNING) skipping? I think it might lead to many bad side-effects like losing your data (i.e. the preferences)

Actually, why this feature can be disabled? what is the use case?

Reply:
This is not implemented by all the services. Only s4l will have it (for now).
This hook is always called, for all the services, before they are closed.

When they get shut down you will see this message. But I guess it's better to log it at warning and see it everywhere than have it silently not show up.

Reply:
Or another thing to do is change the name of the function to something like `attempt_save_user_services_preferences`